### ISSUE TITLE
Add support for BigVolumeViewer

### ISSUE THREAD

**Issue by constantinpape** [mobie/mobie-viewer-fiji#415]
Add support for scenery / BigVolumeViewer (to potentially replace the current 3d viewer?!).
This can be done via BDV-playground, relevant links:
https://github.com/bigdataviewer/bigdataviewer-playground/tree/master/src/main/java/sc/fiji/bdvpg/bvv
https://github.com/bigdataviewer/bigdataviewer-playground/tree/master/src/main/java/sc/fiji/bdvpg/scijava/command/bvv

This was first discussed in https://github.com/mobie/mobie-viewer-fiji/issues/237.

**Comment by ekatrukha** [comment]
Hello @tischi et al.,

I've started to play around with it. For now, I am stuck with just running it.
I am using  [OpenPlatybrowser.java](https://github.com/mobie/mobie-viewer-fiji/blob/main/src/test/java/examples/OpenPlatybrowser.java) as a test, but it gives me an error 
`libblosc.so: cannot open shared object file`
I guess I need to install it? Or should I just add to my debug JVM 
something like `-Djna.library.path=/..Fiji.app/lib/linux64/lib`? Gonna try that

**Comment by ekatrukha** [comment]
Ok, I figured it out

**Comment by tischi** [comment]
> Ok, I figured it out

Actually, I am curious, what did you need to do?

**Comment by tischi** [comment]
...and @ekatrukha the web address is wrong in the OpenPlatyBrowser.java.

Please use: `https://github.com/mobie/platybrowser-project`

**Comment by ekatrukha** [comment]
Here is a [first attempt](https://github.com/mobie/mobie-viewer-fiji/compare/main...ekatrukha:mobie-viewer-fiji:main) on using BVV to render 3D images (that works).
I should have made another branch, gonna correct this later.
For now I use original BVV without extra things.

I use your trick to convert a specific resolution of the source to RAI, but it should be possible 
to use Source directly.
Somehow when I try it directly, it hangs for a long time. 
I suspect BVV tries to load everything, all levels, if the data does not come as 'nice spimdata'.
How would you show this list of `SourceAndConverter` in Bdv normally?
I am going to look/dig into it later.

**Comment by tischi** [comment]
> Somehow when I try it directly, it hangs for a long time.

That is very strange, I would think the having the multiple resolutions should help...

`BdvFunctions.show( ... )` has an implementation with a `List< SourceAndConverter >` as input argument. Does `BvvFunctions.show()` have that, too? If not, maybe we can ask Tobias to add it...?

**Comment by ekatrukha** [comment]
No, not yet, it is more fun to figure it out.

**Comment by tischi** [comment]
And I think we should for now not replace the old way of showing volumes but rather add another `[ ] BV` checkbox next to the existing `[ ] V` checkbox. Is that something you could do? Otherwise I can make a branch where I do it and then you could continue from there. 

Another consideration is that I already have a `scijava39` branch. We could consider using this as a base for our developments. I am not sure it matters though...what do you think?

**Comment by ekatrukha** [comment]
For now, I am going to try experimenting and writing things in a clean way, so it is not important.
Append transforms and deal with sources addition and removal (gonna use logic from 3D viewer).
I think an additional BV checkbox would be better. 

Some extra questions, does Mobie work with 32bit/float data? 

My preliminary plan is to make things work in 'classic' `bvv`, then switch to `bvv-playground` that is a bit more advanced (LUTs/volumetric render), should be easy. Another bonus is that playground should work with pom 38/39 without problems.
I also thought to make a `bvv-playground-minimal` with a minimal GUI for clipping/scale/zoom and maybe leave an animation panel. What do you think about it?
I mean, once something is ready, you can look at it, test and decide.

**Comment by tischi** [comment]
`bvv-playground-minimal`: Would you then depend on this yourself? If so, then yes!

MoBIE: Float data: Yes, for instance the X-Ray tomography data that we deal with is float.

Everything else: Sounds awesome! Thanks so much! This is exciting!!

**Comment by ekatrukha** [comment]
Ok, I finally figured out, I think, what was wrong. Your `Source` seems always have a time point, even if it is a single volume. BVV loads them differently than BDV, it needs to be wrapped up in `spimdata` in this case. Otherwise it first loads everything to the memory. Time option is tricky.

Btw, if it is a 'true' time data, should we load just a single time point or allow user to change it?
Not sure what will happen if there are multiple sources with different number of time points.

This wrapping also should take care of the 8-bit data, since BVV natively shows only 16-bit. And in principle, float too. Maybe you can point me in the direction of the float data?

This `bvv-minimal` yes, should be handy, I am going to reuse in a number of our projects too. But it will take a month or so.

**Comment by tischi** [comment]
Good question about the `time`.... I think we did not have volumetric time-lapse data as a use-case yet.
In general, of course, it would be good to link the time-sliders of BDV and BVV, I guess using some listener pattern.

BTW: this reminds me that Nico Chiaruttini shared a link with me how he is doing this in bdv-playground: 

https://github.com/search?q=repo%3Abigdataviewer%2Fbigdataviewer-playground%20bvv&type=code

Probably it would be useful to see how he does it....

I will look for publicly accessible float data.

**Comment by ekatrukha** [comment]
Yes, I saw and tried the way Nicolas is doing it, but it 
A) in principle works only with 16bit sources and 
B) does not work with sources supplied by Mobie. BVV does not recognize them as volatile multires, so it loads everything as Simple Stack, i.e. reads all data into memory.

After a couple of days reading code I am coming to the conclusion that to show them in BVV via `spimdata` wrap is a good solution, but then it is re-caching. I will try to finish this implementation till the end of the week and see.

Spimdata seems to provide control over cache, which is important for BVV, probably because of the way it uploads things to GPU cache.

While 'regular' caching (via `cache-examples`) does not expose cache control explicitly (or I could not find it) and probably cache access is organized differently. But BDV is less picky or does its own caching differently, since it is just one plane.

It is exciting problem to work on. I thought it will be super easy and quick, but here we go. But I learned a lot about 'internals' behind.

Nicolas definitely should know a lot about it, he does reflection cache substitutions and other things I do not understand yet in BDV playground.
But I will contact him next week, since I did not hit a wall yet. And (mainly) because at that time I will finish some other project that I promised to him 🙈

Probably he knows how to expose cachecontrols, so no need to recache.

Worse case scenario, instead of RGB copy of resulution level loaded to 3DViewer we can load 'raw' RAI to BVV.

**Comment by tischi** [comment]
Hi @ekatrukha,

Nice! Thanks for all the exploration! In the end, if wrapping into SpimData works, why not. Maybe it is even useful that it is re-caching there, because for the 3-D display very different data will need to be loaded than for the 2-D display?

In general though, I have no clue about BVV, but I am very surprised that one cannot simply add a SourceAndConverter to it, which, as you probably know,  exposes the cached `volatile` version of the data via `sourceAndConverter.asVolatile()`. Thus I would hope that at a simple `BvvFunction.show( sac )` should do the trick? I think we should discuss this with Nico and Tobi. Maybe via a Zulip thread? 

Regarding the 16-bit: We could do something like

`
        RandomAccessibleInterval< UnsignedShortType > shortTypes =
                Converters.convert( floatTypes, ( floatType, shortType )
                    -> shortType.setReal( scale * ( floatType.get() - min ) ), new UnsignedShortType() );
`

We would then have to do this for the whole sourceAndConverter, i.e. all resolution levels, both volatile and non-volatile. But before implemeting this it would be good to check with Nico, because chances are that he has something for this already in bdv-playground.

**Comment by tischi** [comment]
Hi,

Here's a link to a float image: 

`https://s3.embl.de/imatrec/ATH_20to200_20240703_PM_01_epo_02_P3_32bit.ome.zarr`

Metadata of highest resolution is:

```
> zarr_overview('https://s3.embl.de/imatrec/ATH_20to200_20240703_PM_01_epo_02_P3_32bit.ome.zarr/0')
Type: Array
Path: https://s3.embl.de/imatrec/ATH_20to200_20240703_PM_01_epo_02_P3_32bit.ome.zarr/0/
Shape: 407 x 4096 x 4096
Chunk Shape: 128 x 128 x 128
No. of Chunks: 4096 (4 x 32 x 32)
Data Type: float32
Endianness: little
Compressor: blosc
```

**Comment by ekatrukha** [comment]
Hello @tischi,

yoohoo, the first milestone, BVV version works, take a look at this [branch](https://github.com/ekatrukha/mobie-viewer-fiji/tree/bvv-classic).
It works with 16-bit and 8-bit images, but `float` should be also ok, providing some `min` and `max`.
Do you do this estimation somewhere? Or we can do iteration over the whole volume to find them,
maybe not so fast, but efficient.

In this branch, I am overriding the same checkbox as for 3DViewer. If you make a branch with a new extra checkbox, I can plug the code there. 

The next step is to include `bvv-playground` for volumetric rendering and LUTs, it should be straightforward since all interfaces are the same, just change package names.
Then `bvv-minimal`, but that will take some weeks.

At some point I was doing debug for 16-bit images using [DebugAngelikaOMEZarrIssue](https://github.com/mobie/mobie-viewer-fiji/blob/fd7fcff2b31e7c8eb1882a27c2acd3208bf1374b/src/test/java/debug/DebugAngelikaOMEZarrIssue.java#L6), but I think something is wrong with that source/transforms. Since it is jumping both in BVV and BDV.

**Comment by ekatrukha** [comment]
I've added BVV Manager to handle additions of different sources without BVV restart and I also figured checkbox thing (I see you also added it in the new branch, thank youuuu).

Now the only thing that is left is 
1) Float data.
2) For Platy project I tried opening segmentation and raw data volumes in BVV and somehow raw is very small, I guess I am missing some scaling factor somewhere, gonna look at it later.

For now, just for convenience, I will still be working [on this branch](https://github.com/ekatrukha/mobie-viewer-fiji/tree/bvv-classic) in my fork. At least until the top two parts are finished.
And when it is ready, I will move it/add it to the desired location.

**Comment by tischi** [comment]
Dear @ekatrukha,

Thanks a lot again! I am on a conference this week. But I will try to squeeze in a quick test! Could you please update the **bvv** branch? I think it will be easier.

**Comment by tischi** [comment]
I fixed the `bvv` branch to compile with `mobie-io 3.0.4` and pushed the changes. 
Can you try with that?

**Comment by tischi** [comment]
Works!

<img width="1393" alt="image" src="https://github.com/user-attachments/assets/e71bddbd-54b9-4196-b8ea-a3ed12687588">


Questions:

1. Do you also support a rendering mode that would look like in the ImageJ 3D Viewer?
2. Is there a way to rotate around the mouse pointer position rather than (0,0,0)? 
3. There is no menu in BVV, I was looking for a "Help" menu that could have the keyboard shortcuts.

**Comment by tischi** [comment]
Do you also get that error (it does not seem to affect anything):

```
Exception in thread "PainterThread" com.jogamp.opengl.GLException: java.lang.InterruptedException
	at com.jogamp.opengl.awt.GLJPanel.display(GLJPanel.java:465)
	at bvv.core.InteractiveGLDisplayCanvas.display(InteractiveGLDisplayCanvas.java:320)
	at bvv.core.VolumeViewerPanel.paint(VolumeViewerPanel.java:517)
	at bdv.viewer.render.PainterThread.run(PainterThread.java:87)
Caused by: java.lang.InterruptedException
	at java.lang.Object.wait(Native Method)
	at java.lang.Object.wait(Object.java:502)
	at java.awt.EventQueue.invokeAndWait(EventQueue.java:1343)
	at java.awt.EventQueue.invokeAndWait(EventQueue.java:1324)
	at com.jogamp.opengl.awt.GLJPanel.display(GLJPanel.java:463)
	... 3 more
```

**Comment by ekatrukha** [comment]
Niiice.

1) "Volumetric" rendering is supported in `bvv-playground`. Check it out, I've made a new `bvvpg` branch. Shortcut `O` (letter) switched between "volume" and max projection. Also in the sources brightness/color dialog mouse right click on the color square allows to select/apply LUT.

2) and 3. No, I do not think there is a mode to rotate differently right now, but it can be implemented, if you have ideas. The shortcuts are almost all [the same as BDV](https://imagej.net/plugins/bdv/#basic-navigation). So arrows to zoom in/out and rotate. `X`,`Y` and `Z` to specify rotation axis, etc.  The list of keymaps is available via `Ctrl + ,` (comma).

Yeah, that error I also have, not sure how to shut down BVV properly. I tried many things, but I think we need to ask Tobias. Indeed, it does not affect things, but it is not nice.

**Comment by tischi** [comment]
Trying `bvvpg`; the `0` key does not do anything for me.

Is that message related?

```
Keymap list file /Users/tischer/Library/Application Support/sc.fiji.bigvolumeviewer/keymaps/keymaps.yaml not found. Using builtin styles.
Appearance settings file /Users/tischer/Library/Application Support/sc.fiji.bigvolumeviewer/appearance.yaml not found. Using defaults.
```

**Comment by ekatrukha** [comment]
No, the message is not related. Do you try `0` or `O`? :) It should be letter O, like in the beginning of Open.

**Comment by tischi** [comment]
Sorry, my bad O works 🥳 

<img width="912" alt="image" src="https://github.com/user-attachments/assets/76839a2a-f51d-4e48-829c-4d3fcee9c1f9">

----

Where exactly should I right click to change the LUT?

**Comment by tischi** [comment]
<img width="912" alt="image" src="https://github.com/user-attachments/assets/21befd45-7531-495c-b138-60dda4316ca3">

Are those strip patterns some rendering artefact? It looks cool anyway!

**Comment by tischi** [comment]
When showing two sources and pressing `O` only one of them is volume rendered; it behaves like a toggle switch...

<img width="1148" alt="image" src="https://github.com/user-attachments/assets/f76a347e-e882-4015-bfc8-25162fd5928d">

Is that on purpose or maybe a bug?

**Comment by ekatrukha** [comment]
You need to click [on this square](https://github.com/UU-cellbiology/bvv-playground?tab=readme-ov-file#lookup-tables-luts). You can also do the same at the Cards panel (shortcut `P`) for groups and etc.

Those patterns yeah, I am not sure, but they add "lighting" atmosphere :)  I think it is probably an artifact of current raytracing with an adaptive step. I need to check that.

**Comment by tischi** [comment]
How hard would it be to add a Glasbey LUT for the label mask images?

**Comment by ekatrukha** [comment]
So it is a good catch with different sources being rendered differently. I've just pushed some update to this branch to have them all rendered consistently. In principle, `bvv-playground` allows mixing them, but it is rarely needed and confusing.

You can use any LUT that is installed in your ImageJ/FIJI if you run it from ImageJ/FIJI. 
If you run it from the IDE, it picks only default LUTs supplied in the `ij.jar`.

**Comment by tischi** [comment]
Very cool!

<img width="912" alt="image" src="https://github.com/user-attachments/assets/044fa33b-21ec-4455-a17b-ddc64fc5774e">

-----

I guess you don't have anything for mesh rendering, do you? This is what would be needed for the current implementation of the rendering of a `SegmentationDisplay`.

To explore the current behaviour you have to add the `sbem-segmentation` `cells` to the view and then `Ctrl + Left Click` in BDV to select one segment and then click the `V` checkbox:

<img width="1449" alt="image" src="https://github.com/user-attachments/assets/c070372f-d14c-43f6-ab0b-c6019c3d0ec4">

That is, only "selected" segments will be 3-D rendered as meshes.

**Comment by ekatrukha** [comment]
I have everything for mesh rendering :) This is how [it happens in BigTrace](https://github.com/ekatrukha/BigTrace/wiki/Panel-View-Clip#roi-surface), for example.  

Do you have already generated meshes? If yes, in which object do you store them, [imagej-mesh](https://github.com/imagej/imagej-mesh)?
Or do you generate them by marching cubes from masks somewhere?

It should not be a problem. The problem is time, I need to finish some other projects first.
This can be a good addition to `bvv-minimal` version with clipping and animation. I thought to call it `BVVBrowser`.

**Comment by tischi** [comment]
Wow, that would be amazing!

The mesh code is all here:

https://github.com/mobie/mobie-viewer-fiji/tree/main/src/main/java/org/embl/mobie/lib/volume

Here is where a specific mesh is added to the volume viewer:

https://github.com/mobie/mobie-viewer-fiji/blob/5d95facf26350d278b787049fe2480f3cc7f3090/src/main/java/org/embl/mobie/lib/volume/SegmentVolumeViewer.java#L283

From a practical point of view, maybe we can create a separate issue for the segment mesh rendering such that we could close this issue here when we released the "normal" BVV rendering. 

And, again, thanks a lot for working on any of this at all! This is amazing and, from my side, there is no rush for anything. I am just super grateful to have you on board!

**Comment by ekatrukha** [comment]
Going back to your remark:

> Is there a way to rotate around the mouse pointer position rather than (0,0,0)?

That is an important point.

It **does** rotate around the mouse pointer position (when you first click the left mouse button).
The same way as in BDV.

I guess the difference with 3DViewer is that the last always applies rotation around "center" (of the universe/current view transform), when you drag the mouse, independently on where the first mouse click happened.
Practically, this results in the situation when the volumes tend to stay "on the canvas".
Do you find it more intuitive?

While in BVV the rotation depends on where you first clicked on the canvas.  
If you first click close to the center of the canvas, it will behave similar to 3DViewer.
If you click close to (any) corner of a window, it would be a new rotation center. 
And it would be far from the current "universe/viewTransform" center. In this case, large mouse movement will make more probable the volume to leave the canvas.

I guess it can be confusing if you are used to 3DViewer navigation approach.

In BigTrace I handle it with extra "centering" [features/options](https://github.com/ekatrukha/BigTrace/wiki/Display-and-navigation#zooming-inout-center). I plan to include it in the upcoming  `BVVBrowser`.
But we can implement 3DViewer navigation style in BVV.
Also, if you have better ideas on this topic, let me know. 
I think navigation of 3D data is challenging and we can try different things and see what works best.

**Comment by ekatrukha** [comment]
In principle, we think about and implement some "Navigation presets" schemes in `bvv-playground`, like,
choose your style: "BDV" or "3DViewer" or "Blender" or something else, depending on what you are familiar with.

For translation, "BDV" uses wheel click, while "3DViewer" uses Shift+wheel click, but it is all a matter of preferences, I guess.

**Comment by tischi** [comment]
## Meshes

https://github.com/mobie/mobie-viewer-fiji/issues/1186

---- 

## Navigation

As I understand it, the centre of rotation is where you click in the current canvas. I do not find this very intuitive for 3D, because this location can be, along the current viewing axis, very far away from the volume that one is looking at.

Could we add something where, e.g. `Ctrl + Left Drag` is rotating around the centre of the "current source"?

----

## BVV Settings

There is the Settings cock wheel that the user can click on for each displayed source in MoBIE. For intensity images that leads to ` ConfigureImageRenderingCommand.class`, this is where we could add some of the BVV settings.

**Comment by ekatrukha** [comment]
**Navigation**

I've added "[3DViewer style rotation](https://github.com/mobie/mobie-viewer-fiji/commit/86f0a032de0afb729eedb523083576147e7e120d)" to the BVV window. 
In addition, shift+left mouse drag is fast, and ctrl+left mouse drag is slow rotation.

**Comment by ekatrukha** [comment]
**Shipping BVV with Mobie**

to FIJI update site you would need to add three jars:
1) [bvv-playground 0.2.9](https://maven.scijava.org/#nexus-search;quick~bvv-playground),
2) [jide-oss](https://mvnrepository.com/artifact/com.formdev/jide-oss/3.7.12) 
3) [flat-laf for it](https://mvnrepository.com/artifact/com.formdev/flatlaf-jide-oss/2.0)

**Comment by tischi** [comment]
Awesome!

Before shipping, I would wait a couple of weeks to see whether we can get also some label mask rendering working, because then we can announce both in one go. If that turns out to be a show-stopper we could decide to only ship the volume rendering.

**Comment by ekatrukha** [comment]
So I think one last thing before the final minimal BVV 'alpha' version, in my opinion is BVV settings.

Some of them can be changed 'on-the-fly' (projection matrix, camera position) and some (render quality, GPU cache size) will require BVV restart and reloading of data.

You mentioned that it can be done at the 'source settings wheel', if I remember correctly.

Are there maybe some 'general Mobie settings' somewhere?

It is just that BVV settings are kind of global (like projection matrix details), not sure it is convenient to add them to the each source settings.
What do you think?

**Comment by tischi** [comment]
Good point. 

I think the easiest (and many only option) right now would be to add another Command here:
https://github.com/mobie/mobie-viewer-fiji/tree/main/src/main/java/org/embl/mobie/command/context

This Command could be then be made accessible to the user as another entry in the Context menu that one sees upon right-click in the BDV window. We could call it "BigVolumeViewer Settings". 

I can add a skeleton of this for you to the bvvpg branch, shall I?

**Comment by ekatrukha** [comment]
Yes, please, that would be very helpful.

**Comment by tischi** [comment]
The question would be where to store those global settings....

I think the easiest right now would be to make all the fields in `BVVManager` `public static`. 

Would that be OK for now?

**Comment by tischi** [comment]
...then, if we want changes to be immediate, once the user changed any of those values we would need to do something like 
```
MoBIE.getInstance().getViewManager().bvvManager.get().coloringChanged()
```


 In the `coloringChanged()` method you would need to add code that ensures that all the relevant settings from BVVManager are applied.

Or, if those changes are too massive, maybe we actually need a new `ImageBVVViewer.bvvSettingsChanged()` method that reinitialises the current viewer more substantially than the current `coloringChanged()`.

**Comment by tischi** [comment]
...and could you please give me one example of a setting that you would like to put there?

**Comment by ekatrukha** [comment]
Hello @tischi,

yes, we can make them `static`. 
But if a user changes them, will they be the same after Mobie restarts?
That would be preferable. I usually store these kind of things using `ij.Prefs.set()` and `ij.Prefs.get()`.
I do not know how it is organized in Mobie.
I mean, if a user has 10 Gb GPU and he wants to change `maxCacheSizeInMB`, 
it should be possible to do it once.

I am in favor of `ImageBVVViewer.bvvSettingsChanged()`.
There we can decide: 
1) if changed parameters only those that [can be changed during runtime](https://github.com/mobie/mobie-viewer-fiji/blob/33c94749a893e878fd1dbf4eb5eeb969f3a3fae8/src/main/java/org/embl/mobie/lib/bvv/BVVManager.java#L19), we do _small update_.
For example `dCam`, `dClipNear` and `dClipFar` can be update 'on-the-fly' via simple `bvv.getBvvHandle().getViewerPanel().setCamParams( ... )` .
2) If other parameters change, we have to close BVV and re-add sources again, it would be _large update_.

**Comment by ekatrukha** [comment]
BTW, I've caught this annoying exception upon closing of BVV, I need to add it to a new version of `bvvpg`.

I've talked to Tobias and indeed, as you mentioned, `show(final SourceAndConverter< T > source )` is absent there, but can be easily added in a manner similar to `bdv`.

It needs to be tested, I will work on it soon. 
It will prevent re-caching of `UnsignedShort` volumes (at least). 
For the other data types we still have to wrap it in 'spimData', so the code that was written is still useful.

Cheers,
Eugene

**Comment by ekatrukha** [comment]
Do you have somewhere 'in the cloud' `UnsignedShort` dataset to test straightforward `SourceAndConverter` addition to `bvv`?

**Comment by tischi** [comment]
I added `test.debug.DebugOpenOpenOrganelle.class`. If you run this you should have a uint16 image from Janelia's OpenOrganelle to play with :-)

Note that you'll also have to rebuild the project, because I merged the latest version of the main branch, which depends on a newer version of mobie-io.

**Comment by ekatrukha** [comment]
Thank youuu! I will work on it next week and keep you updated!

**Comment by tischi** [comment]
I forgot to mention one thing: Since `ConfigureBVVRenderingCommand.class implements Command` the parameter values persist, I think also when closing and re-opening Fiji (or the IDE). That's functionality one get's for free by the SciJava Command framework. Thus, we will not need `ij.Prefs.set()`.

However, what is missing is that we would automatically apply the recent settings of `ConfigureBVVRenderingCommand` to BVV upon startup of MoBIE. I don't know how to fetch those values without actually running the Command. I will create a new issue for this and ask Curtis.

**Comment by ekatrukha** [comment]
I've implemented parameters that do not require BVV restart,
gonna see now how I can restart BVV.
One way would be to kill it and reload all `sac` s,
but I guess it is not very efficient.
I think I would need to unload `BvvSource` 's somewhere and load them back.

**Comment by ekatrukha** [comment]
I have a trouble.
To restart it, in the `ConfigureBVVRenderingCommand` [here](https://github.com/mobie/mobie-viewer-fiji/blob/bvvpg/src/main/java/org/embl/mobie/command/context/ConfigureBVVRenderingCommand.java) I need access to the `imageBVViewer`, so I can reload the list of added `SourceAndConverter` s.
Do you know how to do it?
Or should we add a link to the `imageBVViewer` in the `BVVManager` class?

**Comment by tischi** [comment]
Does this work?

```
Bvv bvv = MoBIE.getInstance().getViewManager().getBvvManager().get();
List< SourceAndConverter< ? > > sources = bvv.getBvvHandle().getViewerPanel().state().getSources();	
```

**Comment by ekatrukha** [comment]
Since it seems to be harder than I though, I moved the discussion on BVV restart to a new issue.

**Comment by ekatrukha** [comment]
Hello @tischi,

In my opinion, we are done with a minimal implementation, what do you think?
I've added BVV view sync with the current BDV shortcut (**D**).
In principle, we can add an option to do it continuously (I think Nicolas does it in `bdvpg`),
but I am not sure it is needed. 
We can wait and see if someone will ask for it.

**Comment by ekatrukha** [comment]
**Shipping BVV with Mobie**

to FIJI update site you would need to add three jars:
1) [bvv-playground 0.3.3](https://maven.scijava.org/service/local/repositories/releases/content/nl/uu/science/cellbiology/bvv-playground/0.3.3/bvv-playground-0.3.3.jar),
2) [jide-oss](https://mvnrepository.com/artifact/com.formdev/jide-oss/3.7.12) 
3) [flat-laf for it](https://mvnrepository.com/artifact/com.formdev/flatlaf-jide-oss/2.0)

**Comment by tischi** [comment]
<img width="249" alt="image" src="https://github.com/user-attachments/assets/d043c5c8-85b5-4b4c-b779-bbc35ef61e5e" />

I guess I should also upload jide-oss ?

**Comment by tischi** [comment]
@ekatrukha I hope I uploaded everything could you try?

**Comment by ekatrukha** [comment]
I've tried on a fresh Fiji install and it works great!
I was able to run it on our server now, with a high bandwidth network connection
and 16 Gb GPU.
It is pretty smooth experience, compared to the laptop!
Yoohoo!

**Comment by tischi** [comment]
Awesome! I will close this thread now in favour of smaller issues.

**Comment by ekatrukha** [comment]
Hello @tischi 

and happy new 2025!

Should we merge `bvvpg` branch to the `main`?

**Comment by tischi** [comment]
Hi, 

Happy new year!

It is merged, afaik...
